llrt_core/src/modules/http/url_search_params.rs

```diff
@@ -1,217 +1,390 @@
-        }
+    pub fn keys(&mut self) -> Vec<String> {
+        self.url
+            .borrow()
+            .query_pairs()
+            .map(|(k, _)| k.to_string())
+            .collect()
     }
 
-    pub fn to_string(&self) -> String {
-        let length = self.params.len();
-        if length == 0 {
-            return String::from("");
+    pub fn set(&mut self, key: Coerced<String>, value: Coerced<String>) {
+        let key = key.0;
+        let value = value.0;
+
+        // Use a HashSet just to filter duplicates
+        let mut uniques = HashSet::new();
+        let mut new_query_pairs: Vec<(String, String)> = Vec::new();
+
+        for (k, v) in self.url.borrow().query_pairs() {
+            // Update the value for an existing key
+            let value = if k == key {
+                value.clone()
+            } else {
+                v.to_string()
+            };
+
+            let query_pair = (k.to_string(), value);
+            if uniques.insert(query_pair.clone()) {
+                new_query_pairs.push(query_pair);
+            }
         }
 
-        fn escape(value: &str) -> String {
-            url::form_urlencoded::byte_serialize(value.as_bytes()).collect()
+        // Append a new key/value pair
+        let query_pair = (key, value);
+        if uniques.insert(query_pair.clone()) {
+            new_query_pairs.push(query_pair);
         }
 
-        let mut string = String::with_capacity(self.params.len() * 2);
-        for (i, (key, value)) in self.params.iter().enumerate() {
-            string.push_str(&escape(key));
-            if !value.is_empty() {
-                string.push('=');
-                string.push_str(&escape(value));
-            }
+        self.url
+            .borrow_mut()
+            .query_pairs_mut()
+            .clear()
+            .extend_pairs(new_query_pairs);
+    }
 
-            if i < length - 1 {
-                string.push('&');
-            }
+    pub fn sort(&mut self) {
+        let mut new_pairs: Vec<(String, String)> =
+            self.url.borrow().query_pairs().into_owned().collect();
+        new_pairs.sort_by(|(a, _), (b, _)| a.cmp(b));
+
+        if new_pairs.is_empty() {
+            self.url.borrow_mut().set_query(None);
+        } else {
+            self.url
+                .borrow_mut()
+                .query_pairs_mut()
+                .clear()
+                .extend_pairs(new_pairs);
         }
-        string
     }
 
-    pub fn keys(&mut self) -> Vec<String> {
-        self.params.iter().map(|(k, _)| k.clone()).collect()
+    pub fn to_string(&self) -> String {
+        // The Url create doesn't properly encode query params for all edge
+        // cases, so we need to construct the query string by percent-encoding
+        // each key/value
+        // TODO: This should probably be fixed in the Url crate
+        self.url
+            .borrow()
+            .query_pairs()
+            .fold(String::new(), |mut acc, (key, value)| {
```

Comment:
Tiny optimization here would be to init string with the capacity of borrow().query().len()*1.25 or something

Reply:
Hmm, interesting idea. I looked through `Url` and `query()` basically unwraps a couple of `Option`s to get offsets to grab the string slice, so should be much faster to check the `len()` than a bunch of reallocations, so makes sense to optimize here. I'm not sure about a multiplier though -- `query()` should already include `=` and `&`, so the `len()` produced by `query()` should be exactly the same as what we're producing here, except for the cases where they aren't properly percent-encoding, in which case we would have some reallocation happening with the `String`. `query()` also returns an `Option<&str>` so we'll  have to unwrap and provide a default as well. I updated it, so let me know what you think. We can revert it or modify if needed.